client.ts

```diff
@@ -29,29 +26,31 @@ export class Client {
 
   async end(): Promise<void> {
     await this._connection.end();
-    delete this._connection;
   }
 
   // Support `using` module
   _aenter = this.connect;
   _aexit = this.end;
 }
 
-// TODO(bartlomieju) to be refactored
-export class PooledClient extends Client {
-  constructor(connection: Connection, release: () => void) {
-    super();
+export class PoolClient {
+  protected _connection: Connection;
+  private _releaseCallback: () => void;
+
+  constructor(connection: Connection, releaseCallback: () => void) {
     this._connection = connection;
-    this.release = function() {
-      release();
-      delete this._connection;
-      delete this.release;
-    };
+    this._releaseCallback = releaseCallback;
   }
 
-  async connect(): Promise<void> {}
+  async query(
+    text: string | QueryConfig,
+    ...args: any[]
+  ): Promise<QueryResult> {
+    const query = new Query(text, ...args);
+    return await this._connection.query(query);
+  }
 
-  async end(): Promise<void> {
-    this.release();
+  async release(): Promise<void> {
+    await this._releaseCallback();
```

Comment:
Isn't this now unsafe if called twice? (puts the connection back into the pool array twice).

Reply:
Good catch @hayd - since you were working on original implementation would you mind taking a look at #49 ?